emacsPackages.tree-sitter: fix grammar/native executor

pimeys opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Fixes missing grammars and adds the Rust core for the emacs tree-sitter package. Now this works, but there are some things I'd like to do better: We download the grammar binaries from the repo. We should maybe use the grammars as we do in the neovim package. This is a bit problematic due to us needing the bundle file available and not all of the grammars are compatible with GPL, meaning some of them such as the json grammar will not load at all.

Anyhow this PR makes it possible to use the tree-sitter in an emacs configuration managed with nix.

Also, I'm still learning nix, so I'm super happy for some guidance how to make this better!

Closes: https://github.com/NixOS/nixpkgs/issues/108089

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
emacsPackages.tree-sitter: fix grammar/native executor

taku0 commented:
LGTM.

```
nix-build -I nixpkgs=. -o emacs -E "let pkgs = (import <nixpkgs> {}); in
  pkgs.emacs.pkgs.withPackages (epkgs: (with epkgs; [
    tree-sitter tree-sitter-langs tsc
  ]))"
./emacs/bin/emacs -q --batch --eval "(progn
  (require 'tree-sitter)
  (require 'tree-sitter-langs)
  (insert \"public class Foo {}\") 
  (java-mode) 
  (tree-sitter-mode) 
  (print (tsc-tree-to-sexp tree-sitter-tree)))"
```

prints:

```
"(program (class_declaration (modifiers) name: (identifier) body: (class_body)))"
```

adisbladis commented on pkgs/applications/editors/emacs/elisp-packages/recipes-archive-melpa.json:
```diff
@@ -106852,35 +106852,6 @@
    "sha256": "1dfmlmmf13qzzfpmpc6lgwjhz8hnz5ys2fw4w3rrz6swfaa2vr53"
   }
  },
- {
```
It looks like this is manually removed. You shouldn't manually remove anything since this file is auto-generated and it will be re-added on the next run.

The manually packaged derivations take precedence over the auto-generated ones anyway so this is essentially a no-op.

pimeys commented:
Good to know. I still at some point want to get the compilation work here and extend the support for other platforms than Linux/x86, but I think this is already quite useful as-is for Linux users who want to build their Emacs configuration by themselves.

pimeys commented:
Trying to use the `compile` script in the langs repo, it requires git and the repository to be in a correct shape. What we basically then need is to use `fetchGit` so that it keeps the `.git` directory:

```nix
src = fetchgit {
  url = "https://github.com/emacs-tree-sitter/tree-sitter-langs.git";
  rev = version;
  sha256 = "sha256-XQU73dzh/QhiOPBWE+wmiE2Yk4HnhM5b4dOVwfX9oxM=";
  fetchSubmodules = true;
};
```

We can evaluate `./script/compile all`, but before the build it runs `git submodule status --cached repos/nix` to all the language repos, and for some reason our `fetchgit` call leaves the repo in a dirty state, so that it removes all the files and adds them again. Is there a way to get a clean checkout with `fetchgit` that doesn't mess around with the repository?

mcwitt commented:
Thanks for working on this! It's working for me on NixOS after updating the elisp-tree-sitter [hash](https://github.com/mcwitt/nixpkgs/commit/8addacafd5eb5163ad546d0e3aad78146111dbf0). Also, should the [queries](https://github.com/emacs-tree-sitter/tree-sitter-langs/tree/master/queries) subdir be included in the package? (`tree-sitter-hl-mode` had no effect for me until I added [this](https://github.com/mcwitt/nixpkgs/commit/b50f18e6a13b5136aded7ed06a5e675d369f196f)).

mcwitt commented on pkgs/applications/editors/emacs/elisp-packages/tree-sitter-langs/default.nix:
```diff
@@ -0,0 +1,54 @@
+  recipe = writeText "recipe" ''
+    (tree-sitter-langs
+    :repo "emacs-tree-sitter/tree-sitter-langs"
+    :fetcher github)
```
I think this is necessary to use the syntax highlighting defaults?
```suggestion
    :fetcher github
    :files (:defaults "queries"))
```

pimeys replied:
Oh yay, I kind of lived without this so far, thanks for pointing it out.

pimeys commented:
> Thanks for working on this! It's working for me on NixOS after updating the elisp-tree-sitter [hash](https://github.com/mcwitt/nixpkgs/commit/8addacafd5eb5163ad546d0e3aad78146111dbf0). Also, should the [queries](https://github.com/emacs-tree-sitter/tree-sitter-langs/tree/master/queries?rgh-link-date=2022-01-16T03%3A22%3A49Z) subdir be included in the package? (`tree-sitter-hl-mode` had no effect for me until I added [this](https://github.com/mcwitt/nixpkgs/commit/b50f18e6a13b5136aded7ed06a5e675d369f196f)).

Thanks for pointing these out. There is something weird in this package, because it works for me on certain computers with the hashes, but then in other systems I get the hash errors. Also updates for now here are super tricky. I need to redo the versions and hashes to the nix scripts multiple times so the package download logic doesn't get triggered. It's probably something to do with writing the bundle version...

It would be really cool to actually compile all the grammars and not relying to the binaries.

mcwitt commented on pkgs/applications/editors/emacs/elisp-packages/tsc/default.nix:
```diff
@@ -0,0 +1,74 @@
+let
+  version = "0.16.1";
+
+  meta = {
```
This is unused. Should it be moved into `symlinkJoin { .. }` below?

mcwitt commented on pkgs/applications/editors/emacs/elisp-packages/tsc/default.nix:
```diff
@@ -0,0 +1,74 @@
+    inherit src;
+
+    pname = "tsc-dyn";
+    commit = version;
```
I might be mistaken but I don't think `buildRustPackage` uses a `commit` arg? (here is the [definition](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/build-rust-package/default.nix))

mcwitt commented:
> Is there a way to get a clean checkout with fetchgit that doesn't mess around with the repository?

Maybe you've already found this, but part of the issue seems to be that `fetchgit` by default doesn't keep the `.git` directory, so the `git submodule` commands run by [tree-sitter-langs-build.el](https://github.com/emacs-tree-sitter/tree-sitter-langs/blob/master/tree-sitter-langs-build.el) will fail.

Apparently there's a way to keep `.git` around by passing `leaveDotGit = true` to `fetchgit`. Although it seems like it's generally [discouraged](https://discourse.nixos.org/t/keep-git-folder-in-when-fetching-a-git-repo/8590/9), that might be a way forward here?

mcwitt commented on pkgs/applications/editors/emacs/elisp-packages/tree-sitter-langs/default.nix:
```diff
@@ -0,0 +1,55 @@
+      install -m444 * $out/langs/bin
+    '';
+  };
+in melpaBuild rec {
```
nit: `rec` not needed here
```suggestion
in melpaBuild {
```

mcwitt commented on pkgs/applications/editors/emacs/elisp-packages/tsc/default.nix:
```diff
@@ -0,0 +1,73 @@
+    sha256 = "sha256-tAohHdAsy/HTFFPSNOo0UyrdolH8h0KF2ekFXuLltBE=";
+  };
+
+  tsc = melpaBuild rec {
```
nit: `rec` not needed
```suggestion
  tsc = melpaBuild {
```

erikbackman commented:
Any updates?

pimeys commented:
I'm using this daily from my own fork. It is just a bit tricky to maintain, so I don't know would it make sense to merge it like this to master.

Would be awesome if somebody could make this to compile all the needed code and skip the downloading logic of the library. Especially version updates are annoyingly tricky to get right.

c0deaddict commented:
> I'm using this daily from my own fork. It is just a bit tricky to maintain, so I don't know would it make sense to merge it like this to master.

Can you elaborate on why it's difficult to maintain? I just started using your changes, and bumped the version to `0.11.3` for both `tsc` and `tree-sitter-grammars` and it just works :tm: (Nice work btw :+1:)

> 
> Would be awesome if somebody could make this to compile all the needed code and skip the downloading logic of the library. Especially version updates are annoyingly tricky to get right.

You mean compiling each of these submodules: https://github.com/emacs-tree-sitter/tree-sitter-langs/tree/master/repos ? I don't think that is a good idea because you'd need a `cargoSha256` for each module, right? That would make it harder to update, although maybe this can be automated with a script of sorts?

bhankas commented:
Paging @adisbladis for another round of review, and @ryantm to check if there is a possibility of automating git submodule updates via the bot. Updates between submodules are sparse enough (lowest 1 day, average ~7days between updates) that it shouldn't be too much of a burden either.

ryantm commented:
If you add a passtrhu.updateScript, the bot will start using it.

https://nixos.org/manual/nixpkgs/stable/#special-variables

adisbladis committed:
emacs.pkgs.tsc: Remove `with` from top-level scope

This is a huge anti-pattern that makes it next to impossible to know
which attrset is providing which attr and it's very easy to make mistakes.

Case in point: This package didn't properly scope `lib.maintainers`.

adisbladis committed:
emacs.pkgs.tsc/tree-sitter-langs: Remove `rec` from derivations

These are not required.

adisbladis committed:
emacs.pkgs.tsc: Add update script

adisbladis committed:
emacs.pkgs.tsc: 0.16.1 -> 0.18.0

Use auto-updater JSON

adisbladis committed:
emacs.pkgs.tree-sitter-langs: Use grammars from nixpkgs and sources from melpa

adisbladis commented:
I have done quite a bit of updates here, the last commit is of particular interest I think (the one using nixpkgs for grammars):
- Minor touchups such as removing `with` and `rec`
- Adding an update script for `tsc`
- Use `tree-sitter-langs` from Melpa Stable
- And the big one: Use grammars from Nixpkgs rather than the precompiled upstream ones

The big one here is the more questionable one, but also what I think might be the most powerful one since upstream doesn't have support for `aarch64`, or it seems, user customizable grammars.
It's also not in the spirit of Nixpkgs to download binaries where we have an option to not do so, practical reasons aside.

For now this is just a static list in the expression, but this could potentially be something that a user of a package would add through some means like:
``` nix
emacs.pkgs.withPackages(epkgs: [
  epkgs.tree-sitter-langs.withGrammars(grammars: [
    grammars.tree-sitter-foo
  ])
])
```

I've only been doing some very basic testing since I'm not already a tree-sitter user. It would be fantastic if someone who is could test all of my changes to this PR and report back.

adisbladis committed:
emacs.pkgs.tree-sitter-langs: Create script to keep grammars up to date with upstream defaults

adisbladis committed:
emacs.pkgs.tree-sitter-langs: Make language plugins configurable

By using an expression like:
``` nix
emacs.pkgs.tree-sitter-langs.passthru.withPlugins (p: [ p.tree-sitter-agda ])
```

adisbladis commented:
I've now also made it possible to customize which languages are linked by:
``` nix
emacs.pkgs.tree-sitter-langs.passthru.withPlugins (p: [ p.tree-sitter-agda ])
```

DamienCassou commented:
Thank you very much @pimeys and everyone involved for your hard work on this one.
